### PR TITLE
Hw4

### PR DESCRIPTION
**Pull request by ErmolinFedor**
Steps + Jenkins

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -48,75 +55,96 @@
...
+                        <groupId>org.aspectj</groupId>
+                        <artifactId>aspectjweaver</artifactId>
+                        <version>${aspectj.version}</version>
+                    </dependency>
+                </dependencies>
+            </plugin>
+            <plugin>
+                <groupId>io.qameta.allure</groupId>
+                <artifactId>allure-maven</artifactId>
+                <version>2.10.0</version>
```
Comment (shoggy) [inline pom.xml]: `${allure-maven.version}`

File: pom.xml [hunk 2]
```diff
@@ -48,75 +55,96 @@
...
+            <artifactId>allure-testng</artifactId>
+            <version>${allure.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.aspectj</groupId>
+            <artifactId>aspectjweaver</artifactId>
+            <version>${aspectjweaver.version}</version>
         </dependency>
```
Comment (shoggy) [inline pom.xml]: I think this dependency does not need to be declared here because it's already declared in `plugins` section

File: pom.xml [hunk 3]
```diff
@@ -48,75 +55,96 @@
...
+                    <dependency>
+                        <groupId>org.aspectj</groupId>
+                        <artifactId>aspectjweaver</artifactId>
+                        <version>${aspectj.version}</version>
+                    </dependency>
+                </dependencies>
+            </plugin>
+            <plugin>
+                <groupId>io.qameta.allure</groupId>
+                <artifactId>allure-maven</artifactId>
```
Comment (shoggy) [inline pom.xml]: The declaration is redundant, it's already defined in `reporting` section and is not used out of reporting scope

File: src/test/java/hw4/base/BaseTests.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+import static org.testng.Assert.assertTrue;
+
+public class BaseTests {
+    protected WebDriver driver;
+    protected Properties property = new Properties();
+    SoftAssert softAssert;
+
+    @BeforeClass
+    public void setUp() {
+        System.setProperty("webdriver.chrome.driver", "D:/chromedriver_win32/chromedriver.exe");
```
Comment (shoggy) [inline src/test/java/hw4/base/BaseTests.java]: Hardcoding paths is not good practice, some outer places better suits this purpose (.properties file, system variable injection), or io.github.bonigarcia:webdrivermanager will be a good solution

File: src/test/java/hw4/base/BaseTests.java [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        try {
+            property.load(
+                    new InputStreamReader(
+                            new FileInputStream("src/test/resources/hw3config.properties"),
+                            Charset.forName("UTF-8")));
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Comment (shoggy) [inline src/test/java/hw4/base/BaseTests.java:50]: hint: `java.util.resourcebundle` can be used

File: src/test/java/hw4/base/TestListener.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import org.testng.ITestResult;
+
+public class TestListener implements ITestListener {
+
+    @Attachment(value = "Page screenshot", type = "image/png")
+    public byte[] saveScreenShot(WebDriver driver) {
+        return ((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES);
+    }
+
+    private static String getTestMethodName(ITestResult itestResult) {
```
Comment (shoggy) [inline src/test/java/hw4/base/TestListener.java]: unused method

File: src/test/java/hw4/base/TestListener.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        return itestResult.getMethod().getConstructorOrMethod().getName();
```
Comment (shoggy) [inline src/test/java/hw4/base/TestListener.java]: I'm not sure, but `itestResult.getName()` could produce same result

File: src/test/java/hw4/pages/components/RightSideBar.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.HashMap;
+import java.util.List;
+
+public class RightSideBar {
+    private WebDriver driver;
+
+    @FindBy(name = "log-sidebar")
+    private WebElement logSidebar;
+
+    @FindBy(css = "[class='panel-body-list logs'] li")
```
Comment (shoggy) [inline src/test/java/hw4/pages/components/RightSideBar.java]: I think this can be simplified `.panel-body-list.logs li`

**Review by shoggy — CHANGES_REQUESTED** [review]

File: src/test/java/hw4/base/TestListener.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @Override
+    public void onTestFailure(ITestResult result) {
+        WebDriver driver = ((BaseTests)result.getInstance()).getDriver();
+        if (driver instanceof WebDriver) {
```
Comment (shoggy) [inline src/test/java/hw4/base/TestListener.java]: If execution flow passes through this line then it is guaranteed that driver is an instance of WebDriver, so you can only compare with null

File: pom.xml (hunk 1 above)
Comment (ErmolinFedor) [reply pom.xml]: fixed

File: pom.xml (hunk 2 above)
Comment (ErmolinFedor) [reply pom.xml]: has removed

File: pom.xml (hunk 3 above)
Comment (ErmolinFedor) [reply pom.xml]: has removed

File: src/test/java/hw4/base/BaseTests.java (hunk 4 above)
Comment (ErmolinFedor) [reply src/test/java/hw4/base/BaseTests.java]: rewritten using the library bonigarcia.wdm

File: src/test/java/hw4/base/BaseTests.java (hunk 5 above)
Comment (ErmolinFedor) [reply src/test/java/hw4/base/BaseTests.java:50]: rewrite some methods using java.util.resourcebundle

File: src/test/java/hw4/base/TestListener.java (hunk 6 above)
Comment (ErmolinFedor) [reply src/test/java/hw4/base/TestListener.java]: has removed

File: src/test/java/hw4/pages/components/RightSideBar.java (hunk 8 above)
Comment (ErmolinFedor) [reply src/test/java/hw4/pages/components/RightSideBar.java]: fixed

File: src/test/java/hw4/base/TestListener.java (hunk 9 above)
Comment (ErmolinFedor) [reply src/test/java/hw4/base/TestListener.java]: fixed

**Review by shoggy — APPROVED** [review]
Looks good

**Pull request merged**